flink-runtime/src/main/java/org/apache/flink/runtime/state/heap/StateTable.java

```diff
@@ -172,14 +229,110 @@ public boolean isEmpty() {
 	 * @return the state of the mapping with the specified key/namespace composite key, or {@code null}
 	 * if no mapping for the specified key is found.
 	 */
-	public abstract S get(K key, N namespace);
+	public S get(K key, N namespace) {
+		int keyGroup = KeyGroupRangeAssignment.assignToKeyGroup(key, keyContext.getNumberOfKeyGroups());
+		return get(key, keyGroup, namespace);
+	}
+
+	public Stream<K> getKeys(N namespace) {
+		return Arrays.stream(state)
+			.flatMap(stateMap -> StreamSupport.stream(Spliterators.spliteratorUnknownSize(stateMap.iterator(), 0), false))
+			.filter(entry -> entry.getNamespace().equals(namespace))
+			.map(StateEntry::getKey);
+	}
+
+	public StateIncrementalVisitor<K, N, S> getStateIncrementalVisitor(int recommendedMaxNumberOfReturnedRecords) {
+		return new StateEntryIterator(recommendedMaxNumberOfReturnedRecords);
+	}
 
-	public abstract Stream<K> getKeys(N namespace);
+	// ------------------------------------------------------------------------
 
-	public abstract StateIncrementalVisitor<K, N, S> getStateIncrementalVisitor(int recommendedMaxNumberOfReturnedRecords);
+	private S get(K key, int keyGroupIndex, N namespace) {
+		checkKeyNamespacePreconditions(key, namespace);
+
+		StateMap<K, N, S> stateMap = getMapForKeyGroup(keyGroupIndex);
+
+		if (stateMap == null) {
```

Comment:
I think this check can go away with the changes that make all maps non-null.

Reply:
This check is to pass the UT of StateBackendTestBase#testKeyGroupSnapshotRestore. The UT will get a key from the backend, but the key is in a key group which not belongs to this backend.

Reply:
I think that there is also a problem with the test because it is attempting to access a key on a backend that is not from a key-group that belongs to the backend. In production this would indicate a bug, so maybe it is better to have this as a precondition check on the backends and throw `IllegalStateException` and adjust the test accordingly.